### PR TITLE
PyGRB rewrite

### PR DESCRIPTION
**Pull request by idorrington92**
pycbc_multi_inspiral is now a coherent matched filter executable. It works correctly (though more tests need to be done) and is integrated into the PyGRB workflow. The workflow crashes at the post-processing stage, as the new coherent matched filter executable produces hdf files but the post-processing still expects xml. We are working on this now. Once this is done, there are relatively minor changes planned, but these should only be to the pycbc_multi_inspiral script.

### CODE REVIEW

**Comment by spxiwh** [comment]
Hi @idorrington92 it's great to see this work becoming ready to merge back onto master! I think the first step to merging this is that the pull request needs rebasing with master as we're seeing changes in 120 files or so.

**Comment by idorrington92** [comment]
Hi Ian,
I've done this now. Started working on the other conflicts.

Thanks,
Iain

**Comment by spxiwh** [comment]
Great, thanks. I've assigned this to both @a-r-williamson and myself to look over (both must approve before merging). You'll have to look over the codeclimate/codacy/codebeat reports and resolve issues where possible (some of the messages are not always so helpful (e.g. cyclomatic complexity) which I don't care about), but check locally with @duncanmmacleod if some of the output here is not clear.

I'll try and look over this properly in the next week or two (deadlines next week)

File: bin/pycbc_multi_inspiral [hunk 1]
```diff
@@ -86,89 +110,254 @@ psd.insert_psd_option_group_multi_ifo(parser)
 scheme.insert_processing_option_group(parser)
 fft.insert_fft_option_group(parser)
 
+#Adding these to see if it lets me use the gating options
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: This comment above seems like it should be removed

File: bin/pycbc_multi_inspiral [hunk 2]
```diff
@@ -86,89 +110,254 @@ psd.insert_psd_option_group_multi_ifo(parser)
...
+from pycbc.vetoes.sgchisq import SingleDetSGChisq
+pycbc.opt.insert_optimization_option_group(parser)
+pycbc.weave.insert_weave_option_group(parser)
+pycbc.inject.insert_injfilterrejector_option_group(parser)
+SingleDetSGChisq.insert_option_group(parser)
 opt = parser.parse_args()
 
+generate_histograms = False
+generate_snr_plots = False
+generate_chi2_plots = False
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: These flags, and the large blocks of plotting at the bottom, look like debugging. Please remove that.

File: bin/pycbc_multi_inspiral [hunk 3]
```diff
@@ -86,89 +110,254 @@ psd.insert_psd_option_group_multi_ifo(parser)
...
+# Polarisation doesn't matter for coherent matched filter so set to zero.
+pol = 0
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: It's confusing to declare polarization here without context. Don't declare `pol=0` here. Instead just put it directly in the call to antenna patters and add the comment.

File: bin/pycbc_multi_inspiral [hunk 4]
```diff
@@ -86,89 +110,254 @@ psd.insert_psd_option_group_multi_ifo(parser)
...
+# Use the time in the middle of the segment to calculate the antenna patterns.
+t_gps = (opt.trig_end_time[0] + opt.trig_start_time[0])/2
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: Why this time? Shouldn't we use the trigger time for this?

File: bin/pycbc_multi_inspiral [hunk 5]
```diff
@@ -86,89 +110,254 @@ psd.insert_psd_option_group_multi_ifo(parser)
...
-    log_level = logging.DEBUG
-else:
-    log_level = logging.WARN
+log_level = logging.DEBUG if opt.verbose else logging.INFO
 logging.basicConfig(format='%(asctime)s : %(message)s', level=log_level)
 
 ctx = scheme.from_cli(opt)
 
+def get_weighted_antenna_patterns(Fp_dict,Fc_dict,sigma_dict):
+    """
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: The function documentation here and below doesn't follow the numpy format that we use in PyCBC.

I'm not going to insist on this while these functions are in a executable, but at some point these will make sense to be moved to a `grb_utils.py` module or similar, and then this *will* need to be fixed.

File: bin/pycbc_multi_inspiral [hunk 6]
```diff
@@ -86,89 +110,254 @@ psd.insert_psd_option_group_multi_ifo(parser)
...
+           sigma_dict: Sigma dictionary for each ifo (sensitivity of each ifo)
+    """
+    #Need the keys to be in alphabetical order
+    keys = sorted(sigma_dict.keys())
+    wp = np.array([sigma_dict[ifo]*Fp_dict[ifo] for ifo in keys])
+    wc = np.array([sigma_dict[ifo]*Fc_dict[ifo] for ifo in keys])
+    return wp,wc
+
+def get_projection_matrix(wp,wc):
+    """
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: *This* function is why you don't need to worry about polarization ... (Not a review request, just a comment).

File: bin/pycbc_multi_inspiral [hunk 7]
```diff
@@ -86,89 +110,254 @@ psd.insert_psd_option_group_multi_ifo(parser)
...
-            try:
-                assert(sample_rate == strain_dict[ifo].sample_rate)
-                assert(flen == strain_segments_dict[ifo].freq_len)
-                assert(tlen == strain_segments_dict[ifo].time_len)
-                assert(delta_f == strain_segments_dict[ifo].delta_f)
-            except:
-                err_msg = "Sample rate, frequency length and time length "
-                err_msg += "must all be consistent across ifos."
-                raise ValueError(err_msg)
+	    sample_rate = strain_dict[ifo].sample_rate
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: No tabs in PyCBC code!

File: bin/pycbc_multi_inspiral [hunk 8]
```diff
@@ -86,89 +110,254 @@ psd.insert_psd_option_group_multi_ifo(parser)
...
 
     logging.info("Making frequency-domain data segments")
-    segments = {}
-    for ifo in opt.instruments:
-        segments[ifo] = strain_segments_dict[ifo].fourier_segments()
-        del strain_segments_dict[ifo]
-
+    segments = {ifo : strain_segments_dict[ifo].fourier_segments() for ifo in opt.instruments}
+    del strain_segments_dict
+    """
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: Please remove, rather than comment, unnecessary code. (Use the `#` character when commenting code).

File: bin/pycbc_multi_inspiral [hunk 9]
```diff
@@ -177,112 +366,229 @@ with ctx:
     # The existing SingleDetBankVeto can calculate the single detector
     # bank veto for multiple ifos, so we just use it directly.
     bank_chisq = vetoes.SingleDetBankVeto(opt.bank_veto_bank_file, flen,
-                                          delta_f, flow, complex64,
-                                          phase_order=opt.order,
-                                          approximant=opt.approximant)
+					  delta_f, flow, complex64,
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: No tabs!

File: bin/pycbc_multi_inspiral [hunk 10]
```diff
@@ -177,112 +366,229 @@ with ctx:
...
-        'chisq'          : float32,
-        'chisq_dof'      : int,
-        'bank_chisq'     : float32,
-        'bank_chisq_dof' : int,
-        'cont_chisq'     : float32
-                }
-    out_vals = {
-        'time_index'     : None,
-        'ifo'            : None,
+	'chisq'          : float32,
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: No tabs!

File: bin/pycbc_multi_inspiral [hunk 11]
```diff
@@ -177,112 +366,229 @@ with ctx:
...
+        'null_snr'      : None,
+        'nifo'          : None
                }
-    names = sorted(out_vals.keys())
+    network_names = sorted(network_out_vals.keys())
 
-    event_mgr = events.EventManagerMultiDet(opt, opt.instruments, names,
-                                        [out_types[n] for n in names], psd=psds)
+    event_mgr = events.EventManagerCoherent(opt, opt.instruments, ifo_names,
+					[ifo_out_types[n] for n in ifo_names], network_names, [network_out_types[n] for n in network_names])
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: No tabs!

File: bin/pycbc_multi_inspiral [hunk 12]
```diff
@@ -177,112 +366,229 @@ with ctx:
...
+            logging.info("Analyzing segment %d/%d" % (s_num + 1, len(segments[ifo]))) 
+            #Find detector sensitivities (sigma) and make array of normalised 
+            sigmasq = {ifo : template.sigmasq(segments[ifo][s_num].psd) for ifo in opt.instruments}
+            sigma = {ifo : np.sqrt(sigmasq[ifo]) for ifo in opt.instruments}
+            stilde = {ifo : segments[ifo][s_num] for ifo in opt.instruments}
+            #Every time s_num is zero, we run new template to increment the template index
+            if s_num==0: event_mgr.new_template(tmplt=template.params, sigmasq=sigmasq)
+            snrv_dict = {}
+            norm_dict = {}
+            corr_dict = {}
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: Indentation seems strange here??

File: bin/pycbc_multi_inspiral [hunk 13]
```diff
@@ -177,112 +366,229 @@ with ctx:
...
+                #					    "snr", cluster_window, ifo)
+ 
         event_mgr.finalize_template_events()
+event_mgr.write_events(opt.output)
+
 
+logging.info("Time to complete analysis: %s" % str(time.time() - time_init))
+
+"""
+#Leaving this code around as a basis for when we implement chisq thresholds.
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: Then please comment with `#`

File: bin/pycbc_multi_inspiral [hunk 14]
```diff
@@ -291,5 +597,97 @@ if opt.fftw_output_float_wisdom_file:
 
 if opt.fftw_output_double_wisdom_file:
     fft.fftw.export_double_wisdom_to_filename(opt.fftw_output_double_wisdom_file)
-
+"""
 logging.info("Finished")
+
```
Comment (spxiwh) [inline bin/pycbc_multi_inspiral]: Remove all the stuff below

File: bin/pygrb/pycbc_make_offline_grb_workflow [hunk 15]
```diff
@@ -177,6 +190,7 @@ if wflow.cp.has_option("workflow-condition_strain", "do-gating"):
     logging.info("Creating gating jobs.")
     wflow.cp = _workflow.set_grb_start_end(wflow.cp, int(sciSegs[ifo][0][0]),
                                            int(sciSegs[ifo][0][1]))
+    """
```
Comment (spxiwh) [inline bin/pygrb/pycbc_make_offline_grb_workflow]: Remove or comment with `#`. Preferably remove.

File: pycbc/events/events.py [hunk 16]
```diff
@@ -136,7 +136,9 @@ def findchirp_cluster_over_window(times, values, window_length):
     indices: Array
         The reduced list of indices of the SNR values
     """
-    assert window_length > 0, 'Clustering window length is not positive'
+    #Commented out the below assert statement to allow for no clustering
+    #in the coherent matched_filter codes
+    #assert window_length > 0, 'Clustering window length is not positive'
```
Comment (spxiwh) [inline pycbc/events/events.py]: Can you revert this change? (Mixing in a change here is going to be problematic with other development work to move to python 3). For now just add a catch in the coherent code to do nothing if window_length is 0.

File: pycbc/events/events.py [hunk 17]
```diff
@@ -337,7 +336,10 @@ def cluster_template_events(self, tcolumn, column, window_size):
         """
         cvec = self.template_events[column]
         tvec = self.template_events[tcolumn]
-        indices = findchirp_cluster_over_window(tvec, cvec, window_size)
+        if window_size == 0:
+            indices = numpy.arange(len(tvec))
```
Comment (spxiwh) [inline pycbc/events/events.py]: This change would probably mean that you don't need the change in `findchirp_cluster` above.

File: pycbc/events/events.py [hunk 18]
```diff
@@ -315,12 +317,9 @@ def keep_loudest_in_interval(self, window, num_keep):
...
         """ Add a vector indexed """
         # initialize with zeros - since vectors can be None, look for the
-        # first one that isn't
+        # longest one that isn't
         new_events = None
-        for v in vectors:
-            if v is not None:
-                new_events = numpy.zeros(len(v), dtype=self.event_dtype)
-                break
+        new_events = numpy.zeros(max([len(v) for v in vectors if v is not None]), dtype=self.event_dtype)
```
Comment (spxiwh) [inline pycbc/events/events.py]: Why was this change necessary?

File: pycbc/events/events.py [hunk 19]
```diff
@@ -546,8 +547,288 @@ def add_template_events_to_ifo(self, ifo, columns, vectors):
...
+        for i, ifo in enumerate(self.ifos):
+            f.prefix = ifo
+            ifo_events = numpy.array([e for e in self.events if \
+                    e['ifo'] == self.ifo_dict[ifo]], dtype=self.event_dtype)
+            if len(ifo_events):
+                ifo_str = ifo.lower()[0] if ifo != 'H1' else ifo.lower()
+                f['snr_%s' % ifo_str] = abs(ifo_events['snr'])
+                f['event_id'] = ifo_events['event_id']
+                try:
+                    # Precessing
```
Comment (spxiwh) [inline pycbc/events/events.py]: Precessing? This approach wouldn't apply directly to a coherent search, unfortunately.

File: pycbc/workflow/core.py [hunk 20]
```diff
@@ -324,13 +324,9 @@ def add_ini_opts(self, cp, sec):
 
                 # Get LFN and PFN
                 for path in values:
-                    # Here I decide if the path is URL or
-                    # IFO:/path/to/file or IFO:url://path/to/file
-                    # That's somewhat tricksy as we used : as delimiter
-                    split_path = path.split(':', 1)
+                    split_path = path.split(':',1)
```
Comment (spxiwh) [inline pycbc/workflow/core.py]: Why did you remove my comment? Also why remove the space that now makes this not PEP8 compliant.

File: pycbc/workflow/core.py [hunk 21]
```diff
@@ -324,13 +324,9 @@ def add_ini_opts(self, cp, sec):
...
                     if len(split_path) == 1:
-                        # Simple case: path to file
-                        ifo = None
+                        ifo=None
```
Comment (spxiwh) [inline pycbc/workflow/core.py]: Again, please revert this change and reinstate the comment.

File: pycbc/workflow/core.py [hunk 22]
```diff
@@ -342,12 +338,17 @@ def add_ini_opts(self, cp, sec):
                             #IFO:path or IFO:URL
                             ifo = split_path[0]
                             path = split_path[1]
+                    print(ifo)
```
Comment (spxiwh) [inline pycbc/workflow/core.py]: We definitely don't want these debugging statements added.

File: pycbc/workflow/core.py [hunk 23]
```diff
@@ -342,12 +338,17 @@ def add_ini_opts(self, cp, sec):
...
+                    print(path)
+                    print(split_path)
+                    print(value)
 
                     curr_lfn = os.path.basename(path)
 
                     # If the file exists make sure to use the
                     # fill path as a file:// URL
                     if os.path.isfile(path):
+                        print("Path is file")
```
Comment (spxiwh) [inline pycbc/workflow/core.py]: Or this one.

File: pycbc/workflow/jobsetup.py [hunk 24]
```diff
@@ -778,6 +779,128 @@ def zero_pad_data_extend(self, job_data_seg, curr_seg):
             new_data_seg = segments.segment([new_data_start, new_data_end])
             return new_data_seg
 
+class PyCBCMultiInspiralExecutable(Executable):
+    """
+    The class responsible for setting up jobs for the
+    pycbc_multi_inspiral executable.
+    """
+    current_retention_level = Executable.MERGED_TRIGGERS
```
Comment (spxiwh) [inline pycbc/workflow/jobsetup.py]: This should take the same value as `PyCBCInspiralExecutable`

File: pycbc/workflow/jobsetup.py [hunk 25]
```diff
@@ -778,6 +779,128 @@ def zero_pad_data_extend(self, job_data_seg, curr_seg):
...
+    #file_input_options = ['--gating-file']
```
Comment (spxiwh) [inline pycbc/workflow/jobsetup.py]: This should now be able to be uncommented.

File: pycbc/workflow/jobsetup.py [hunk 26]
```diff
@@ -778,6 +779,128 @@ def zero_pad_data_extend(self, job_data_seg, curr_seg):
...
+        node.add_opt('--trig-end-time', valid_seg[1])
+
+        node.add_profile('condor', 'request_cpus', self.num_threads)
+
+        # Set the input and output files
+        node.new_output_file_opt(data_seg, '.hdf', '--output',
+                                 tags=tags, store_file=self.retain_files)
+        node.add_input_opt('--bank-file', parent, )
+
+        # TODO: isn't there a cleaner way of doing this?
```
Comment (spxiwh) [inline pycbc/workflow/jobsetup.py]: Shouldn't you be using `--frame-files` like `PyCBCInspiral` ... That might need a big "FIXME" in flashing lights, as this would prevent the code running in places like OSG.

File: pycbc/workflow/matched_filter.py [hunk 27]
```diff
@@ -293,25 +294,27 @@ def setup_matchedfltr_dax_generated_multi(workflow, science_segs, datafind_outs,
...
-        cp.set('inspiral', 'block-duration',
-               str(abs(science_segs[ifos[0]][0]) - \
-                       2 * int(cp.get('inspiral', 'pad-data'))))
+    if match_fltr_exe == 'pycbc_multi_inspiral':
+        exe_class = select_matchedfilter_class(match_fltr_exe)
+        cp.set('inspiral', 'longitude',\
+               str(radians(float(cp.get('workflow', 'ra')))))
+        cp.set('inspiral', 'latitude',\
+               str(radians(float(cp.get('workflow', 'dec')))))
+        # from pycbc.workflow.grb_utils import get_sky_grid_scale
```
Comment (spxiwh) [inline pycbc/workflow/matched_filter.py]: Remove unneeded old code.

File: bin/pycbc_multi_inspiral (hunk 4 above)
Comment (idorrington92) [reply bin/pycbc_multi_inspiral]: For the onsource it makes sense to use the trigger time, but that will cause problems when this is run on the offsource. Perhaps we could somehow set it to have the trigger time when running the on the onsource and use the middle of the segment otherwise?

File: pycbc/workflow/matched_filter.py (hunk 27 above)
Comment (idorrington92) [reply pycbc/workflow/matched_filter.py]: I wasn't sure whether this might be needed when we add in searching over a sky grid? Should I just delete it?

File: pycbc/workflow/jobsetup.py (hunk 26 above)
Comment (idorrington92) [reply pycbc/workflow/jobsetup.py]: I have no idea with this. Maybe @pannarale  can help with this?

File: bin/pycbc_multi_inspiral (hunk 12 above)
Comment (idorrington92) [reply bin/pycbc_multi_inspiral]: Why? 
This is the first time I've ever really had to care about how my code looks but I can't see anything strange here.

File: bin/pycbc_multi_inspiral (hunk 4 above)
Comment (spxiwh) [reply bin/pycbc_multi_inspiral]: I'm a little confused now. In `coh_PTF` the antenna patterns are calculated using the time of the GRB and those same antenna patters used for the offsource ... This does mean that the RA changes, but that's okay, you want to keep the antenna patterns constant to assess the background.

So for the onsource the GRB time should be used, otherwise you will lose sensitivity.

For background the GRB time should also be used otherwise you will be calculating the appropriate background.

File: pycbc/workflow/jobsetup.py (hunk 24 above)
Comment (spxiwh) [reply pycbc/workflow/jobsetup.py]: Yes

File: pycbc/workflow/jobsetup.py (hunk 26 above)
Comment (spxiwh) [reply pycbc/workflow/jobsetup.py]: @pannarale Is there some sort of to-do list for GRB development? Fixing this may not belong in this pull request, but this shouldn't get lost.

File: pycbc/workflow/matched_filter.py (hunk 27 above)
Comment (spxiwh) [reply pycbc/workflow/matched_filter.py]: If this is being left in for a reason, then please indicate that as an in-line comment.

File: bin/pycbc_multi_inspiral (hunk 4 above)
Comment (idorrington92) [reply bin/pycbc_multi_inspiral]: Ok, that makes sense. I'll edit the setup to get the trigger time given to pycbc_multi_inspiral and we'll use that.

File: pycbc/events/events.py (hunk 16 above)
Comment (idorrington92) [reply pycbc/events/events.py]: I reverted this without any other changes because I couldn't remember where it caused problems, and it ran fine, so I think we can just revert it and leave it at that. (Though I assume I got rid of it at some point for a reason so this may cause problems in the future.)

File: pycbc/events/events.py (hunk 17 above)
Comment (idorrington92) [reply pycbc/events/events.py]: Didn't see this before commenting on the previous thing. I agree :p

File: pycbc/events/events.py (hunk 18 above)
Comment (idorrington92) [reply pycbc/events/events.py]: I can't remember. I've reverted it for now, and hopefully it won't break anything later on.

File: pycbc/workflow/core.py (hunk 20 above)
Comment (idorrington92) [reply pycbc/workflow/core.py]: This is just from when you sent me this code through slack, I never updated it to when you fixed it properly. Fixing this now.

File: pycbc/events/events.py (hunk 19 above)
Comment (idorrington92) [reply pycbc/events/events.py]: Should I just delete this then? I just left it in because it was there for the multi det event manager.

File: pycbc/events/events.py (hunk 18 above)
Comment (idorrington92) [reply pycbc/events/events.py]: I remembered why this is necessary. Some of the pycbc multi inspiral outputs are 'None' or only have one value per template/segment. Others have one value per trigger. When these get written to hdf, if there is only one value, then it gets written out repeatedly (once for each trigger). 

I can probably fix this in the executable if you prefer (all the outputs have the same length when it reaches this point). Or we can make the change posted above.
